python/pyspark/sql/connect/session.py

```diff
@@ -215,7 +215,38 @@ def createDataFrame(
         _inferred_schema: Optional[StructType] = None
 
         if isinstance(data, pd.DataFrame):
-            _table = pa.Table.from_pandas(data)
+            from pandas.api.types import (  # type: ignore[attr-defined]
+                is_datetime64_dtype,
+                is_datetime64tz_dtype,
+            )
+            from pyspark.sql.pandas.types import (
+                _check_series_convert_timestamps_internal,
+                _get_local_timezone,
+            )
+
+            # Copying the frame to avoid modifying it.
+            data_copy = data.copy()
+            # We need double conversions for the truncation, first truncate to microseconds.
+            for col in data_copy:
+                if is_datetime64tz_dtype(data_copy[col].dtype):
+                    data_copy[col] = _check_series_convert_timestamps_internal(
+                        data_copy[col], _get_local_timezone()
+                    ).astype("datetime64[us, UTC]")
+                elif is_datetime64_dtype(data_copy[col].dtype):
+                    data_copy[col] = data_copy[col].astype("datetime64[us]")
+
+            # Create a new schema and change the types to the truncated microseconds.
+            pd_schema = pa.Schema.from_pandas(data_copy)
```

Comment:
Hm .. I wonder if we cast the type after creating the Arrow batch here. Calling `pa.Schema.from_pandas` already copies and coverts pandas DataFrame.